README.md
# get-pip-pyopenssl

This tool generates working `get-pip` scripts that force the use of
`pyOpenSSL` inside `pip`. Because it is not possible to install packages
until having a `pip` that actually works, all the required dependencies
are appended to the end of the scripts, as the usual `get-pip` does with
`pip`.

The resulting `get-pip` scripts solve PyPI issues
[#974](https://github.com/pypa/pypi-support/issues/974) and
[#978](https://github.com/pypa/pypi-support/issues/978) that left `pip`
unusable for the Python versions without SNI support.

The project is currently in active development. At the moment, the tool
provides support for Python 2.6 and Python 2.7 under GNU/Linux and
Windows. Future releases will try to add support for more Python
versions.

## Usage

There is one specific script for every platform, architecture and Python
ABI, because the bundled dependencies `cffi` and `cryptography` are
pre-compiled binaries.

The helper script `get-pip-pyopenssl.py` takes care of downloading the
appropriate script under the hood and it is the recommended way of using
the tool.

* For GNU/Linux systems, run:
```sh
wget http://pylegacy.org/hub/get-pip-pyopenssl.py
python get-pip-pyopenssl.py
```

* For Windows, run (in PowerShell):
```sh
Invoke-WebRequest http://pylegacy.org/hub/get-pip-pyopenssl.py -OutFile get-pip-pyopenssl.py
python get-pip-pyopenssl.py
```

## License

```
Copyright (c) 2021-2022 Víctor Molina García

get-pip-pyopenssl is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

get-pip-pyopenssl is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with get-pip-pyopenssl. If not, see <https://www.gnu.org/licenses/>.
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. The format
is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and the
project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.4.0] - 2022-02-04

### Added
- Pipelines to upload distributables to pylegacy site.
- Add `SubjectAltNameWarning` to list of ignored warnings.

### Fixed
- Fix setup encoding comment to deal with corner case under PowerShell.
- Improve code compatibility across Python versions.

## [0.3.0] - 2021-07-30

### Changed
- Move version-specific scripts inside a `pip` folder when building.

## [0.2.0] - 2021-07-23

### Fixed
- Silence expected warnings `SNIMissingWarning` and `InsecurePlatformWarning`
  while still patching `pip`.

## [0.1.0] - 2021-07-18

### Added
- Initial Python 2.6 and Python 2.7 support for the following platforms:
  - GNU/Linux: archs i686 and x86_64, ABIs cp26m, cp26mu, cp27m, cp27mu.
  - Windows: archs win32 and win_amd64, ABIs cp26m, cp27m.
- Helper script `get-pip-pyopenssl.py` that finds the specific script based
  on the system information provided by the Python environment.


[Unreleased]:
https://github.com/molinav/get-pip-pyopenssl/compare/v0.4.0...master
[0.4.0]:
https://github.com/molinav/get-pip-pyopenssl/compare/v0.3.0...v0.4.0
[0.3.0]:
https://github.com/molinav/get-pip-pyopenssl/compare/v0.2.0...v0.3.0
[0.2.0]:
https://github.com/molinav/get-pip-pyopenssl/compare/v0.1.0...v0.2.0
[0.1.0]:
https://github.com/molinav/get-pip-pyopenssl/releases/tag/v0.1.0


scripts/build.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2021-2022 Víctor Molina García
#
# This file is part of get-pip-pyopenssl.
#
# get-pip-pyopenssl is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# get-pip-pyopenssl is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with get-pip-pyopenssl. If not, see <https://www.gnu.org/licenses/>.
#
"""Script to build `get-pip-pyopenssl` distributables."""
from __future__ import print_function


def main():
    """Main script function."""

    import io
    import os
    import re
    import sys
    import argparse
    import itertools
    import subprocess
    from generate import __version__

    # Define arguments.
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--dest",
        type=str, help="Destination build folder", required=False,
        default="build")
    parser.add_argument(
        "--remote",
        type=str, help="Expected remote root location", required=False,
        default=None)

    # Parse arguments.
    args = parser.parse_args()

    here = os.path.dirname(__file__)
    targets = ("Linux", "Windows")
    archs = ("32bit", "64bit")
    abis = ("cp26m", "cp26mu", "cp27m", "cp27mu")

    for target, arch, abi in itertools.product(targets, archs, abis):

        # Do not build 'mu' ABI for Windows.
        if target == "Windows" and abi.endswith("u"):
            continue

        # Get Python version from ABI.
        version = re.match(r"cp(\d+)m?u?", abi).groups(1)[0]
        version = ".".join([version[0], version[1:]])

        # Call the generate script.
        print("- Building {0} for {1} {2}...".format(abi, target, arch))
        subprocess.call([
            sys.executable, "-u",
            os.path.join(here, "generate.py"),
            "--target", target,
            "--arch", arch,
            "--abi", abi,
            "--dest", os.path.join(args.dest, "pip", version),
        ])

    # Write out the helper script.
    template = os.path.join(here, "template-main.py")
    outfile = os.path.join(args.dest, "get-pip-pyopenssl.py")
    with io.open(outfile, "wb") as fd1:
        with io.open(template, "r", encoding="utf-8") as fd2:
            for line2 in fd2:
                if line2 == "__version__ = None\n":
                    line2 = "__version__ = \"{0}\"\n".format(__version__)
                if line2.startswith("    scriptroot =") and args.remote:
                    line2 = "    scriptroot = \"{0}\"\n".format(args.remote)
                fd1.write(line2.encode("utf-8"))


if __name__ == "__main__":
    main()


scripts/template.py
#! /usr/bin/env python
# -*- coding: utf8 -*-
#
# Copyright (c) 2021 Víctor Molina García
#
# This file is part of get-pip-pyopenssl.
#
# get-pip-pyopenssl is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# get-pip-pyopenssl is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with get-pip-pyopenssl. If not, see <https://www.gnu.org/licenses/>.
#
"""Script to install a functional `pip` package that uses `pyOpenSSL`.

This script installs a functional `pip` package for Python versions
without Server Name Identification (SNI) support by forcing the use
of `pyOpenSSL` inside `pip`.

Because it is not possible to install packages until having a `pip`
that actually works, all the required dependencies are appended to
the end of the script, as the usual `get-pip.py` does with `pip`.

The resulting `get-pip.py` script solves PyPI issues #974 and #978
that left `pip` unusable for the Python versions without SNI support:
    https://github.com/pypa/pypi-support/issues/974
    https://github.com/pypa/pypi-support/issues/978
"""
from __future__ import print_function


def unpack(path, dest=None):

    import os
    from contextlib import closing
    from zipfile import ZipFile

    pkgname = os.path.basename(path).split("-")[0]
    if dest is None:
        dest = os.getcwd()
    with closing(ZipFile(path, "r")) as archive:
        for file in archive.namelist():
            if file.startswith("{0}/".format(pkgname)):
                archive.extract(file, dest)


def pkgdecode(text):

    from base64 import b64decode
    return b64decode("".join(line.strip() for line in text.split("\n")))


def pip_extract(pkgname, dest=None):

    import os
    import shutil
    import tempfile

    pkg = PACKAGES[pkgname]
    tmpdir = None
    try:
        tmpdir = tempfile.mkdtemp(prefix="tmp-get-pip-extract-")
        pkgpath = os.path.join(tmpdir, pkg["filename"])
        pkgdata = pkgdecode(pkg["filedata"])
        with open(pkgpath, "wb") as fd:
            fd.write(pkgdata)
        unpack(pkgpath, dest=dest)
    finally:
        if tmpdir:
            shutil.rmtree(tmpdir, ignore_errors=True)


def pip_install(pkgname, *args):

    import sys
    import subprocess
    import pip

    pip_main = lambda *a: pip.main(list(a))
    if hasattr(pip, "_internal"):
        pip_call = [sys.executable, "-m", "pip"]
        pip_main = lambda *a: subprocess.call(pip_call + list(a))

    rc = pip_main("install", pkgname, *args)
    if rc != 0:
        raise RuntimeError("pip failed with exit code {0}".format(rc))


def pip_autoinstall(pkgname, *args):

    import os
    import shutil
    import tempfile

    pkg = PACKAGES[pkgname]
    tmpdir = None
    try:
        tmpdir = tempfile.mkdtemp(prefix="tmp-get-pip-autoinstall-")
        pkgpath = os.path.join(tmpdir, pkg["filename"])
        pkgdata = pkgdecode(pkg["filedata"])
        with open(pkgpath, "wb") as fd:
            fd.write(pkgdata)
        pip_install(pkgpath, *args)
    finally:
        if tmpdir:
            shutil.rmtree(tmpdir, ignore_errors=True)


def pip_autopatch():

    import os
    import pip

    # Define files to patch.
    pip_fold = os.path.dirname(pip.__file__)
    pyopenssl_file = os.path.join(pip_fold, "_vendor", "urllib3",
                                  "contrib", "pyopenssl.py")

    pip_file = os.path.join(pip_fold, "__init__.py")
    compat_file = os.path.join(pip_fold, "_vendor", "distlib", "compat.py")
    sslimport_file = compat_file if os.path.exists(compat_file) else pip_file

    # Force `pip` to use `pyOpenSSL`.
    lines = []
    with open(sslimport_file, "r") as fd:
        found_try = False
        for line in fd:
            if "try:" in line:
                found_try = True
            elif found_try and "import ssl" in line:
                indent = line[:len(line) - len(line.lstrip())]
                injection = [
                    "import warnings",
                    "with warnings.catch_warnings():",
                    "    warnings.simplefilter(\"ignore\", category=DeprecationWarning)",
                    "    from pip._vendor.urllib3.contrib import pyopenssl",
                    "    pyopenssl.inject_into_urllib3()",
                    "    del pyopenssl",
                ]
                lines.extend(["{0}{1}\n".format(indent, item)
                               for item in injection])
                found_try = False
            else:
                found_try = False
            lines.append(line)
    with open(sslimport_file, "w") as fd:
        fd.writelines(lines)

    # Patch issue with unicode/bytes mix in `pyopenssl`.
    lines = []
    with open(pyopenssl_file, "r") as fd:
        text = "return self.connection.send(data)"
        for line in fd:
            lines.append(line.replace("data", "data.encode()")
                         if text in line else line)
    with open(pyopenssl_file, "w") as fd:
        fd.writelines(lines)


def main():

    import os
    import sys
    import imp
    import shutil
    import tempfile

    tmpdir = None
    curdir = os.getcwd()
    force_args = ["-I", "--no-deps"]

    try:

        tmpdir = tempfile.mkdtemp(prefix="tmp-get-pip-")
        os.chdir(tmpdir)

        # Unpack `pip` and `wheel` temporarily.
        for pkg in ("pip", "wheel"):
            pip_extract(pkg)
        sys.path.insert(0, tmpdir)

        # Install `pip`, `wheel` and `setuptools`.
        for pkg in ("pip", "argparse", "wheel", "setuptools"):
            pip_autoinstall(pkg, *force_args)

        # Delete temporary `pip` and `wheel` and reload the installed ones.
        sys.path.pop(0)
        for pkg in ("pip", "wheel"):
            shutil.rmtree(pkg, ignore_errors=True)
            imp.reload(imp.load_module(pkg, *imp.find_module(pkg)))

        # Install `cffi` and its dependencies.
        for pkg in ("pycparser", "cffi"):
            pip_autoinstall(pkg)

        # Install `enum34` and its dependencies.
        for pkg in ("ordereddict", "enum34"):
            if pkg in PACKAGES:
                pip_autoinstall(pkg, *force_args)

        # Install `cryptography` dependencies.
        for pkg in ("six", "asn1crypto", "idna", "ipaddress"):
            pip_autoinstall(pkg)

        # Install `cryptography` and `pyOpenSSL`.
        for pkg in ("cryptography", "pyOpenSSL"):
            pip_autoinstall(pkg)

        # Reload `pip` again and patch it.
        imp.reload(imp.load_module("pip", *imp.find_module("pip")))
        pip_autopatch()
        print("Successfully patched pip")

    finally:

        if tmpdir:
            shutil.rmtree(tmpdir, ignore_errors=True)
        os.chdir(curdir)


PACKAGES = {}


if __name__ == "__main__":
    main()


scripts/template-main.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2021-2022 Víctor Molina García
#
# This file is part of get-pip-pyopenssl.
#
# get-pip-pyopenssl is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# get-pip-pyopenssl is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with get-pip-pyopenssl. If not, see <https://www.gnu.org/licenses/>.
#
"""Script to install a functional `pip` package that uses `pyOpenSSL`.

This is a helper script that downloads the appropriate `get-pip-pyopenssl`
script based on the Python installation used to run it.

The `get-pip-pyopenssl` script will install a functional `pip` package
for Python versions without Server Name Identification (SNI) support by
forcing the use of `pyOpenSSL` inside `pip`, which solves PyPI issues
#974 and #978 that left `pip` unusable for the Python versions without
SNI support:
    https://github.com/pypa/pypi-support/issues/974
    https://github.com/pypa/pypi-support/issues/978
"""

__version__ = None


def get_arch():
    """Return the platform name."""

    import sys
    import distutils.util

    value = distutils.util.get_platform().replace("-", "_")
    if value.startswith("macosx"):
        raise NotImplementedError
    if value == "linux_x86_64" and sys.maxsize == 2147483647:
        value = "linux_i686"
    return value.replace("linux", "manylinux1")


def get_abi():
    """Return the ABI for the current Python installation."""

    import sys
    import platform
    from distutils import sysconfig

    # Get ABI flags.
    abid = ("d" if sysconfig.get_config_var("WITH_PYDEBUG") == 1 or
            hasattr(sys, "gettotalrefcount")
            else "")
    abim = ("m" if sys.version_info < (3, 8) and
            sysconfig.get_config_var("WITH_PYMALLOC") == 1 or
            platform.python_implementation() == "CPython"
            else "")
    abiu = ("u" if sys.version_info < (3, 3) and
            sysconfig.get_config_var("Py_UNICODE_SIZE") == 4 or
            sys.maxunicode == 0x10FFFF
            else "")

    # Create ABI string.
    pyver = "cp{0}{1}".format(*sys.version_info[:2])
    pyabi = "{0}{1}{2}{3}".format(*[pyver, abid, abim, abiu])

    return pyabi


def main():
    """Main script call."""

    import os
    import re
    import sys
    import shutil
    import tempfile
    import subprocess

    try:
        from urllib.request import urlopen
    except ImportError:
        from urllib2 import urlopen

    arch = get_arch()
    pyabi = get_abi()
    pyver = re.match(r"(cp\d+)m?u?", pyabi).groups(1)[0]
    version = "{0}.{1}".format(*sys.version_info[:2])

    scriptroot = os.path.dirname(os.path.abspath(__file__))
    scriptname = "get-pip-pyopenssl-{0}-{1}-{2}.py".format(pyver, pyabi, arch)
    if re.match("https?://.*", scriptroot):
        # Script root is an URL.
        scriptpath = "/".join([scriptroot.strip("/"), "pip", version,
                               scriptname])
        try:
            tmpdir = tempfile.mkdtemp(prefix="tmp-get-pip-pyopenssl-")
            tmppath = os.path.join(tmpdir, scriptname)
            conn = urlopen(scriptpath)
            try:
                with open(tmppath, "wb") as fd:
                    fd.write(conn.read())
            finally:
                conn.close()
            subprocess.call([sys.executable, "-u", tmppath])
        finally:
            if tmpdir:
                shutil.rmtree(tmpdir, ignore_errors=True)
    else:
        # Script root is a folder.
        scriptpath = os.path.join(scriptroot, "pip", version, scriptname)
        subprocess.call([sys.executable, "-u", scriptpath])


if __name__ == "__main__":
    main()


scripts/generate.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2021-2022 Víctor Molina García
#
# This file is part of get-pip-pyopenssl.
#
# get-pip-pyopenssl is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# get-pip-pyopenssl is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with get-pip-pyopenssl. If not, see <https://www.gnu.org/licenses/>.
#
"""Script to create alternative `get-pip.py` scripts that use `pyOpenSSL`.

This script generates working `get-pip.py` scripts for Python versions
without Server Name Identification (SNI) support by forcing the use
of `pyOpenSSL` inside `pip`.

Because it is not possible to install packages until having a `pip`
that actually works, all the required dependencies are appended to
the end of the script, as the usual `get-pip.py` does with `pip`.

The resulting `get-pip.py` script solves PyPI issues #974 and #978
that left `pip` unusable for the Python versions without SNI support:
    https://github.com/pypa/pypi-support/issues/974
    https://github.com/pypa/pypi-support/issues/978
"""

__version__ = "0.4.0+dev"


def makedirs(name, mode=511, exist_ok=False):
    """Create a leaf directory and all intermediate ones.

    It works like `mkdir`, except that any intermediate path segment (not
    just the rightmost) will be created if it does not exist. If the
    target directory already exists, raise an ``OSError`` if ``exist_ok``
    is False. Otherwise no exception is raised. This is recursive.
    """

    import os
    import errno

    exist_ok = bool(exist_ok)
    try:
        os.makedirs(name, mode)
    except OSError as err:
        if err.errno == errno.EEXIST and exist_ok and os.path.isdir(name):
            return
        raise


class cachedproperty(property):  # pylint: disable=invalid-name
    """Property that caches its value after first calculation."""

    def __init__(self, fget=None, fset=None, fdel=None, doc=None):
        """Initialise property and save also the cached property name."""

        super(cachedproperty, self).__init__(fget, fset, fdel, doc)
        self.name = fget.__name__

    def __get__(self, obj, objtype=None):
        """Property getter that handles value caching."""

        if obj is None:
            return self
        if self.fget is None:
            raise AttributeError("unreadable attribute")

        try:
            cache = obj.__cache__
        except AttributeError:
            obj.__cache__ = {}
            cache = obj.__cache__
        try:
            return cache[self.name]
        except KeyError:
            cache[self.name] = self.fget(obj)
            return cache[self.name]


class Package(object):
    """Wrapper class for Python packages coming from PyPI."""

    def __init__(self, filename):
        """Create a new instance from a Python package filename."""

        self.filename = filename
        self.data = None

    @property
    def name(self):
        """Python package name."""

        nsuffixes = 1 + int(self.filename.endswith(".tar.gz"))
        base = self.filename.rsplit(".", nsuffixes)[0]
        return base.split("-")[0]

    @property
    def version(self):
        """Python package version in string format."""

        nsuffixes = 1 + int(self.filename.endswith(".tar.gz"))
        base = self.filename.rsplit(".", nsuffixes)[0]
        return base.split("-")[1]

    @property
    def author(self):
        """Package author as shown in PyPI."""

        import re

        pattern = ".*<p><strong>Author:</strong> <a href=\".*\">(.*)</a></p>"
        for htmlrow in self.pypi_project_html.splitlines():
            match = re.match(pattern, htmlrow)
            if match:
                return match.group(1)
        msg = "no author found for package {0}".format(self.filename)
        raise ValueError(msg)

    @property
    def license(self):
        """Package license as shown in PyPI."""

        import re

        pattern = ".*<p><strong>License:</strong> (.*)</p>"
        for htmlrow in self.pypi_project_html.splitlines():
            match = re.match(pattern, htmlrow)
            if match:
                license = match.group(1)
                if re.match(r"MIT( License( \(UNKNOWN|MIT.*\)?))?", license):
                    license = "MIT License (MIT)"
                elif re.match(r"BSD( License( \(UNKNOWN|BSD.*\)?))?", license):
                    license = "BSD License (BSD)"
                return license
        msg = "no license found for package {0}".format(self.filename)
        raise ValueError(msg)

    @property
    def pypi_project_url(self):
        """PyPI project url in string format (file download view)."""

        urlpattern = "https://pypi.org/project/{0}/{1}/#files"
        return urlpattern.format(self.name, self.version)

    @cachedproperty
    def pypi_project_html(self):
        """PyPI project HTML in string format (file download view)."""

        try:
            from urllib.request import urlopen
        except ImportError:
            from urllib2 import urlopen

        # This is just a temporarily workaround but needs a real fix.
        # pylint: disable=protected-access
        import ssl
        ssl._create_default_https_context = ssl._create_unverified_context

        conn = urlopen(self.pypi_project_url)
        try:
            html = conn.read().decode("utf-8")
        finally:
            conn.close()
        return html

    @property
    def pypi_package_url(self):
        """Python package remote url from the PyPI repository."""

        import re

        filename_regex = self.filename.replace(".", "\\.")
        pattern = ".*<a href=\"(.*{0}.*)\">".format(filename_regex)

        for htmlrow in self.pypi_project_html.splitlines():
            match = re.match(pattern, htmlrow)
            if match:
                return match.group(1)
        msg = "no url found for package {0}".format(self.filename)
        raise ValueError(msg)

    def download(self):
        """Get the Python package as a :class:`bytes` object."""

        try:
            from urllib.request import urlopen
        except ImportError:
            from urllib2 import urlopen

        conn = urlopen(self.pypi_package_url)
        try:
            self.data = conn.read()
        finally:
            conn.close()

    def textify(self, indent=0):
        """Return the Python package data as plain encoded text."""

        if self.data is None:
            self.download()

        return "\n".join([
            "{indent}\"{name}\": {{",
            "{indent}    \"author\":",
            "{indent}        \"{author}\",",
            "{indent}    \"license\":",
            "{indent}        \"{license}\",",
            "{indent}    \"filename\":",
            "{indent}        \"{filename}\",",
            "{indent}    \"filedata\": \"\"\"",
            "{filedata}",
            "{indent}    \"\"\",",
            "{indent}}},"
        ]).format(name=self.name,
                  author=self.author,
                  license=self.license,
                  filename=self.filename,
                  filedata=self.pkgencode(self.data),
                  indent=" " * indent)

    @staticmethod
    def pkgencode(data, pad=0, nchars=None):
        """Return data string from a data stream using base64."""

        spaces = " " * pad
        if nchars is None:
            nchars = 79 - pad

        from base64 import b64encode
        raw = b64encode(data).decode("utf-8")
        lines = [raw[i:i + nchars] for i in range(0, len(raw), nchars)]
        return "\n".join(["{0}{1}".format(spaces, line) for line in lines])

    @staticmethod
    def pkgdecode(text):
        """Return data stream from a data string using base64."""

        from base64 import b64decode
        return b64decode("".join(line.strip() for line in text.split("\n")))


def main():
    """Main script function."""

    import io
    import re
    import os.path
    import argparse

    valid_targets = {
        ("Windows", "32bit"):
            "win32",
        ("Windows", "64bit"):
            "win_amd64",
        ("Linux", "32bit"):
            "manylinux1_i686",
        ("Linux", "64bit"):
            "manylinux1_x86_64",
    }

    # Define arguments.
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--target",
        type=str, help="Target operating system", required=True,
        choices=["Linux", "Windows"],)
    parser.add_argument(
        "--arch",
        type=str, help="Architecture", required=True,
        choices=["32bit", "64bit"])
    parser.add_argument(
        "--abi",
        type=str, help="Python ABI implementation", required=True,
        choices=["cp26m", "cp26mu", "cp27m", "cp27mu"])
    parser.add_argument(
        "--dest",
        type=str, help="Destination folder", required=False,
        default="./")

    # Parse arguments.
    args = parser.parse_args()
    version = re.match(r"(cp\d+)m?u?", args.abi).groups(1)[0]
    semver = ".".join(re.match(r"cp(\d)(\d+)", version).groups())
    label = "-".join([version, args.abi,
                      valid_targets[(args.target, args.arch)]])

    # Do not allow 'mu' implementations for Windows.
    if args.target == "Windows" and args.abi.endswith("mu"):
        msg = "unsupported Python ABI version '{0}' under {1} {2}"
        raise ValueError(msg.format(args.abi, args.target, args.arch))

    if semver == "2.6":
        cffi_version = "1.10.0" if args.target == "Windows" else "1.11.2"
        crypto_version = "2.0.3" if args.target == "Windows" else "2.1.1"
        packages = [
            # Essential packages (`pip`, `wheel` and `setuptools`).
            Package("pip-9.0.3-py2.py3-none-any.whl"),
            Package("argparse-1.4.0-py2.py3-none-any.whl"),
            Package("wheel-0.29.0-py2.py3-none-any.whl"),
            Package("setuptools-36.8.0-py2.py3-none-any.whl"),
            # `cffi` and dependencies (for `cryptography`).
            Package("pycparser-2.18.tar.gz"),
            Package("cffi-{0}-{1}.whl".format(cffi_version, label)),
            # `enum34` and dependencies (for `cryptography`).
            Package("ordereddict-1.1.tar.gz"),
            Package("enum34-1.1.10-py2-none-any.whl"),
            # `cryptography` and its remaining dependencies.
            Package("six-1.13.0-py2.py3-none-any.whl"),
            Package("asn1crypto-1.4.0-py2.py3-none-any.whl"),
            Package("idna-2.7-py2.py3-none-any.whl"),
            Package("ipaddress-1.0.23-py2.py3-none-any.whl"),
            Package("cryptography-{0}-{1}.whl".format(crypto_version, label)),
            # `pyOpenSSL` and its remaining dependencies.
            Package("pyOpenSSL-16.2.0-py2.py3-none-any.whl"),
        ]
    elif semver == "2.7":
        packages = [
            # Essential packages (`pip`, `wheel` and `setuptools`).
            Package("pip-20.3.4-py2.py3-none-any.whl"),
            Package("argparse-1.4.0-py2.py3-none-any.whl"),
            Package("wheel-0.36.2-py2.py3-none-any.whl"),
            Package("setuptools-44.1.1-py2.py3-none-any.whl"),
            # `cffi` and dependencies (for `cryptography`).
            Package("pycparser-2.20-py2.py3-none-any.whl"),
            Package("cffi-1.14.6-{0}.whl".format(label)),
            # `enum34` and dependencies (for `cryptography`).
            Package("enum34-1.1.10-py2-none-any.whl"),
            # `cryptography` and its remaining dependencies.
            Package("six-1.16.0-py2.py3-none-any.whl"),
            Package("asn1crypto-1.4.0-py2.py3-none-any.whl"),
            Package("idna-2.10-py2.py3-none-any.whl"),
            Package("ipaddress-1.0.23-py2.py3-none-any.whl"),
            Package("cryptography-2.2.2-{0}.whl".format(label)),
            # `pyOpenSSL` and its remaining dependencies.
            Package("pyOpenSSL-18.0.0-py2.py3-none-any.whl"),
        ]
    else:
        msg = "unsupported Python ABI version '{0}' under {1} {2}"
        raise ValueError(msg.format(args.abi, args.target, args.arch))

    pkgtext = []
    for pkg in packages:
        pkgtext.append(pkg.textify(indent=4))
    injection = "\n".join(pkgtext)

    here = os.path.dirname(__file__)
    template_file = os.path.join(here, "template-script.py")
    target_name = "get-pip-pyopenssl-{0}.py".format(label)
    makedirs(args.dest, exist_ok=True)
    with io.open(os.path.join(args.dest, target_name), "wb") as fd1:
        with io.open(template_file, "r", encoding="utf-8") as fd2:
            for line2 in fd2:
                if line2 == "#! /usr/bin/env python\n":
                    line2 = "#! /usr/bin/env python{0}\n".format(semver)
                if line2 == "__version__ = None\n":
                    line2 = "__version__ = \"{0}\"\n".format(__version__)
                if line2 == "PACKAGES = {}\n":
                    line2 = "PACKAGES = {{\n\n{0}\n\n}}\n".format(injection)
                fd1.write(line2.encode("utf-8"))


if __name__ == "__main__":
    main()
